internal/policy/container/has_modified_files.go

```diff
@@ -24,140 +28,149 @@ var _ check.Check = &HasModifiedFilesCheck{}
+	// order within the maps.
+	for idx, layer := range layers {
+		layerIDHash, err := layer.DiffID()
 		if err != nil {
-			return nil, fmt.Errorf("could not extract layers: %w", err)
-		}
-		pathChan := make(chan string)
-
-		var wg sync.WaitGroup
-		wg.Add(1)
-		go func() {
-			// For each path in the pathChan, add it to the layer's
-			// list of files.
-			layerFiles := make([]string, 0)
-			for path := range pathChan {
-				layerFiles = append(layerFiles, path)
-			}
-			// also add it to the overall list of files.
-			files = append(files, layerFiles)
-			wg.Done()
-		}()
-		// add paths to the pathChan
-		err = untar(pathChan, r)
+			return nil, nil, fmt.Errorf("unable to retrieve diff id for layer: %w", err)
+		}
+		layerID := layerIDHash.String()
+
+		layerDir := filepath.Join(layerDir, fmt.Sprintf("%02d-%s", idx, layerID))
+		err = fs.Mkdir(layerDir, 0o755)
 		if err != nil {
-			return nil, fmt.Errorf("failed to extract tarball: %w", err)
+			return nil, nil, fmt.Errorf("could not create layer directory: %w", err)
 		}
 
-		wg.Wait() // wait for file list to get appended to the files var
-	}
+		layerIDs = append(layerIDs, layerID)
 
-	files, dropped := p.dropFirstLayerIfEmpty(files)
-	if dropped {
-		// tell the user the first layer was dropped
-		diff0, _ := layers[0].DiffID()
-		diff1, _ := layers[1].DiffID()
-		logger.V(log.DBG).Info(
-			fmt.Sprintf("The first layer (%s) contained no files, so the next layer (%s) is being used as the base layer.",
-				diff0.String(),
-				diff1.String(),
-			))
-	}
+		files, err := generateChangesFor(layer)
+		if err != nil {
+			return nil, nil, err
+		}
 
-	return &packageFilesRef{files, packageFiles}, nil
-}
+		found, pkgList := findRPMDB(ctx, layer)
+		if !found {
+			logger.V(log.TRC).Info("could not find rpm database in layer", "layer", layerID)
+			if idx > 0 {
+				// Just make this is the same as last layer, since the RPM db was not modified
+				lastLayer := layerIDs[idx-1]
+				layerRefs[layerID] = packageFilesRef{
+					LayerFiles:        files,
+					LayerPackages:     layerRefs[lastLayer].LayerPackages,
+					LayerPackageFiles: layerRefs[lastLayer].LayerPackageFiles,
+					HasRPMDB:          false,
+				}
+				continue
+			}
 
-// dropFirstLayerIfEmpty will evaluate the length of the first layer and will drop its entry in files if it's empty.
-// This avoids establishing a baseline of files installed in the original layer with an empty list. An example
-// of when this should occur is in cases where the image was built FROM scratch, such as ubi-micro. This is required
-// to ensure we don't attempt to validate against an empty layer.
-func (p *HasModifiedFilesCheck) dropFirstLayerIfEmpty(files [][]string) ([][]string, bool) {
-	var dropped bool
-	if len(files[0]) == 0 {
-		files = files[1:] // shift the empty layer out.
-		dropped = true
+			// If it's the first layer, just make the pkgList empty.
+			pkgList = make([]*rpmdb.PackageInfo, 0)
+		}
```

Comment:
I'm still not understanding this block. We're checking _every_ layer for an RPMDB? I assume this is because we generate a package list again, make sure that all packages have the right info, etc. I guess my concern is that the log output ends up looking like this.

```
time="2023-03-08T12:16:05-06:00" level=trace msg="findRPMDB found an RPM db" layer="sha256:296e14ee24149e14d573a1fbf5b5a625c7bb0cc22f5b2a8b180e833258187948"
time="2023-03-08T12:16:05-06:00" level=trace msg="could not find rpm database in layer" layer="sha256:8f26704f753cbe8b20afc01e97a4cb652befe26791b4d33193afc99ce544ebf5"
time="2023-03-08T12:16:05-06:00" level=trace msg="could not find rpm database in layer" layer="sha256:96915ad71c5c10538d1f3f5dda36a8d90dc8151805acb3b263406cb9e50036ff"
time="2023-03-08T12:16:05-06:00" level=trace msg="could not find rpm database in layer" layer="sha256:d6f7de083499e5152f183f885248de0ec370972fd9956bf3c8e48091984e750a"
time="2023-03-08T12:16:05-06:00" level=trace msg="could not find rpm database in layer" layer="sha256:1b8d55546b61cab3c1ecccdd581f3a184fb6c8b1916b254a406c6a395b0e880d"
time="2023-03-08T12:16:05-06:00" level=trace msg="could not find rpm database in layer" layer="sha256:e979289ce7d578b603bada6f7eeca5fc9f13b55b43d631e1b9626166ffcc778d"
time="2023-03-08T12:16:05-06:00" level=trace msg="findRPMDB found an RPM db" layer="sha256:63b8c90a3decf1db1660fbf451af06ce62ba843d46a77ce2dc0a0c8faeac9e50"
time="2023-03-08T12:16:05-06:00" level=trace msg="could not find rpm database in layer" layer="sha256:c8ec39fea4bb24e466b6fc2c5919b0d3f670097acb44635768fa388f0bcd9ac9"
time="2023-03-08T12:16:05-06:00" level=trace msg="could not find rpm database in layer" layer="sha256:0f39727bac0d55d024d8524e66a68a621eca212e8a6718f2accb3452ada55506"
time="2023-03-08T12:16:06-06:00" level=trace msg="could not find rpm database in layer" layer="sha256:7eb228728e47c149e57edaf5e822d6814d9be390c1535beabe32369067373390"
time="2023-03-08T12:16:07-06:00" level=trace msg="could not find rpm database in layer" layer="sha256:bc3e54ef33b1f796a97e73040c1323daedf2135ee559369ef3ba726d2b8a9f78"
time="2023-03-08T12:16:07-06:00" level=debug msg="found disallowed modification in layer" file=usr/sbin/nginx layer="sha256:c8ec39fea4bb24e466b6fc2c5919b0d3f670097acb44635768fa388f0bcd9ac9"
time="2023-03-08T12:16:07-06:00" level=debug msg="found disallowed modification in layer" file=usr/sbin/nginx-debug layer="sha256:c8ec39fea4bb24e466b6fc2c5919b0d3f670097acb44635768fa388f0bcd9ac9"
```

What's throwing me is the subsequent RPMDB being found. In trace logs, this doesn't help me debug unless I already know the implementation inside and out. Maybe a nit. I'm not sure.

Reply:
Trace is for developers. Right? I wouldn't want a partner to have to use trace in order to know what failed. It's most certainly an implementation detail. 

But to answer your initial question... Yes. Every layer is being check for an RPM db.